docs/events/glides/com.mbta.ctd.glides.vehicle_trip_assignments.v1.mdx

```diff
@@ -0,0 +1,184 @@
+
+The `vehicleId` should be used for checking for equality and tracking updates. The `carNumbers` MAY change (for example reversing the order, or updating bad data from a bad AVI read) while referring to the same train. Car numbers are included for context and to help match to other systems, but are not guaranteed to exactly match any other data source. `carNumbers` MAY be `null` if the vehicle has been removed from realtime tracking and there are no car numbers associated with it anymore.
+
+Glides SHOULD NOT publish a sequence of events that imply the same car number is associated with two different `vehicleId`s at the same time.
+
+### VehicleTripAssignment
+
+Fields:
+
+- `vehicle` ([Vehicle](#vehicle) | null):
+- `tripKey` ([TripKey](#tripkey) | null):
+
+`null` means "unassigned":
+
+- An assignment that includes both a `tripKey` and a `vehicle` means that the given vehicle is operating the given trip at this point in time.
+- An assignment that includes a `tripKey` but a `null` `vehicle` means that no known vehicle is operating that trip right now.
+- An assignment that includes a `vehicle` but a `null` `tripKey` means that the vehicle is not currently operating a trip.
+- Assignments SHOULD NOT have both fields `null`, but if they do, they can be ignored.
+
+## Updating Assignments
+
+Consumers SHOULD maintain state about which vehicle-trip assignments are active, and update that state when they receive an assignment that refers to a `vehicleId` or `tripKey`.
+
+Vehicles can only be assigned to one trip at a time. A vehicle that was previously assigned to `TRIP A` MAY be assigned directly to `TRIP B` without a separate assignment that assigns its trip to `null` first. Conversely, a trip can only have one assigned vehicle at a time, and doesn't need a previous vehicle to be removed before having a new vehicle assigned to it.
+
+If `VEHICLE 1` was last assigned to `TRIP A`, and a new event assigns `VEHICLE 1` to `TRIP B`, Glides MUST also send another `VehicleTripAssignment` in the same event's batch to assign the `vehicle` for `TRIP A` to `null` or to a different vehicle. Consumers are not expected to deduce that `TRIP A` no longer has a vehicle assigned based on an update to `VEHICLE 1` and `TRIP B` (which doesn't mention `TRIP A`). Glides MUST ensure that after processing all updates in an event's batch, if the last update to `TRIP A` assigned `VEHICLE 1` to it, then `VEHICLE 1` MUST NOT have any other more recent updates.
+
+Conversely, if `VEHICLE 1` was last assigned to `TRIP A`, and a new event assigns `VEHICLE 2` to `TRIP A`, Glides MUST also assign `VEHICLE 1` to another trip or to a `null` trip in the same batch. By the end of processing all of the updates in each batch, if the last update to `VEHICLE 1` assigned it to `TRIP A`, then `TRIP A` MUST NOT have any other more recent updates.
```

Comment:
I ended up with a couple awkward rules here to prevent consumers' state from getting tangled up. I'm happy with the current design, but if anybody finds the state-updating rules to be overly complex, here could be an alternative to discuss:

Instead of assigning vehicles or trips to null, we could do `{vehicle: v, trip: t, type: "assign"}` and `{vehicle: v, trip: t, type: "unassign"}`, and Glides has to unassign a pair before reassigning either of them.

Problems with the current spec:
- It's a bit complex.
- It says that you only end up in a good state after you finish processing the whole batch of assignments.
- If Glides messes up, or while a batch is partially applied, you could have a vehicle assigned to a trip, but the trip not assigned to the vehicle.

This alternative's pros:
- State is valid after every single assignment object, not just after the whole batch
- Removes the nullable fields and a couple edge cases.

This alternative's cons:
- Still requires the producer (Glides) to send updates in just the right order, and not send invalid updates. - - Is a bit backwards-facing (what was the old assignment, so we can undo it), while the current written spec is more forwards-facing (ignore the past, what's the train doing right now?).
- Could still end up in an invalid state if Glides messes up or if a consumer misses an event.

Reply:
It feels like a lot of the complication here is because this schema as-written is trying to handle both vehicle to trip and trip to vehicle assignments, when the main underlying goal is vehicle to trip assignments and I'm not sure that we need to explicitly spell out the reverse direction. If we want clients to be able to deduce the reverse mapping, we can just spell out the semantics to indicate that upon receiving a Vehicle A -> Trip 1 mapping, it should clear out any other Vehicle X -> Trip 1 mappings in its state.

> ...requires the producer (Glides) to send updates in just the right order, and not send invalid updates.

Are there actually any barriers to Glides sending these events in the right order?

> Could still end up in an invalid state if Glides messes up or if a consumer misses an event.

What sort of failure modes would cause either of these things to happen?

Reply:
Overall I'm more of a fan of the alternative, but I think it can be simplified even further:

> Glides has to unassign a pair before reassigning either of them.

Can't we just make the semantics of processing the `{vehicle: v, trip: t, type: "assign"}` message that the consumer has to remove any vehicle -> trip mappings involving either vehicle `v` or trip `t` before inserting the new mapping from `v` to `t`?

Reply:
Oooh, I had never considered that we wouldn't need an explicit "trip -> vehicle" mapping, too. And that implicitly clearing state like that would be fine to ask clients to do. Like, it shifts logic from the producer to the consumer, which I normally don't like, but it does remove some edge cases in the spec, so I think it's worth it.

> Are there actually any barriers to Glides sending these events in the right order?

No, I'm just pointing out that the alternative doesn't eliminate illegal states.

> What sort of failure modes would cause either of these things to happen?

If a Glides bug results in an invalid message (e.g. reassigning a second vehicle to the same trip), or if a client misses a message and gets the state out of sync so that a valid message _looks_ like it's reassigning a second vehicle to the same trip, then the client might not have a good way to handle the event that leaves it in a legal state. (This wouldn't happen at start up from missing old history. It would only happen if a client receives some events, then misses an event, then receives another event.) This is easy to avoid by just not writing any bugs, but I'm pointing it out as a flaw in both the current version and alternative I proposed. A well-designed spec would always have a defined way to handle any incoming message.

Actually, I think your last idea, where both vehicles and trips can both be implicitly unassigned, could end up with no messages that are illegal to handle from any state, and updates can be processed one at a time instead of in batches. This might be the ticket. I'll sleep on it over the weekend and try writing it up on Monday.

Reply:
A followup question: You suggested only vehicle to trip assignments (vehicle is not null, trip might be null). What if it was the other way around, and the feed was trip to vehicle assignments (trip is not null, vehicle is nullable. For each trip, which vehicle is on it?)

This doesn't match as well with RTR's use case, where it has a vehicle and wants to know its trip. But it would better fit with the `trips_updated` feed, where `tripKey` is the id, and car numbers are the data.

Or maybe the schema could be symmetrical where neither one of them needs to be the primary key, either of them could be null, and the client can use either one as a key depending on their use? I think the benefits from this discussion come from the implicit unassignment in either direction, not from focusing on vehicle to trip assignments.

Again, I'll try it out Monday and see how it goes.

Reply:
Looking at the updates I like that it's simplified! I guess we could go either way on the directionality, but I think there's benefit in choosing one direction and sticking with it, just because otherwise if there's an active assignment of Vehicle A to Trip 1, then there are actually two different ways to unassign it with the same semantics (assigning Vehicle A to trip null or assigning vehicle null to Trip 1).

Reply:
Did the implicit unassignments in [00399b4](https://github.com/mbta/schemas/pull/9/commits/00399b406442269fc470c6f9e1aa0fe3e96f3c85). It's definitely better.

I haven't changed the symmetry between vehicles and trips. The main benefit to making it asymmetrical would be that consumers don't need to handle either vehicle being null OR trip being null. But it wouldn't change the state that consumers need to save, or simplify how updates are made. I'm leaning towards leaving it as symmetrical because it makes less promises about non-null fields, so is more forwards compatible.

Reply:
> I'm leaning towards leaving it as symmetrical because it makes less promises about non-null fields, so is more forwards compatible.

How much forward compatibility does this actually buy us? If we allow trip ID to be null, allowing null vehicle IDs with non-null trip IDs just gives us a different way of expressing something that could already be expressed with a non-null vehicle ID and a null trip ID, and if we wanted it to mean something different that would be a breaking change in the semantics anyway.

Reply:
It'd matter if we ever want to give other information about a trip, and want to be able to give that information even when there's no vehicle assigned. If null vehicles are allowed, then we can start publishing `{tripKey: {key}, vehicle: null, foo: "whatever"}` and it would work in a totally backwards-compatible way. If we ban null vehicles, then we have to change the spec and update all consumers before we start publishing this data.

Asymmetric nonnull vehicles would also implies that vehicles-to-trips is the preferred way to view the data. It's the way that RTR would use it, but I don't think it's inherent. After all `trips_updated` uses trips as keys and does trip-to-vehicle matching. Another consumer could use trips as keys for their matching state if that makes sense for their use case.

So it's just hypothetical benefits, but the cost (adding a null case) is small. Keep in mind that RTR will already have to write a `unassignVehiclesFromTrip(state, tripKey)` function anyway, because it could be triggered by `{vehicle: {nonnull}, tripKey: {was assigned to another vehicle}}`. This is just about whether you also need a `{vehicle: null, trip: tripKey} ->` case to call that existing function.

Reply:
> Another consumer could use trips as keys for their matching state if that makes sense for their use case.

Sure, but then they could do that by maintaining a vehicle to trip mapping based on these events and then reversing it.

> It'd matter if we ever want to give other information about a trip, and want to be able to give that information even when there's no vehicle assigned.

Why would we put that other information in a `vehicle_trip_assignment` event as opposed to just creating a new event?

Reply:
Not a strongly held belief, but I find myself leaning towards `vehicle` being non-nullable. The main reason for that is allowing either to be nullable means there's multiple way to accomplish the same thing. If I need to unassign vehicle 1 from trip 2, I can either send `%{vehicle: 1, trip: null}` or `%{vehicle: null, trip: 2}`. Picking vehicle as the key forces us to choose a lane and reduces the needed handling code on the consumer side.

That said, if, for some reason, the implementation is easier on the glides side, if we also support the trip-oriented version as well (or instead!) I'm not inherently opposed. But from the limited amount I've seen, Glides seems somewhat vehicle-oriented (Eddie or Sky, please correct me if I'm off base on that) so having looked at exactly zero code my guess is that the vehicle-oriented version is probably what Glides would end up sending in most situations anyway.